src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,218 @@
+
+            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, newFileNameWithExtension);
+
+            bool errorsDetected = false;
+            foreach (var action in documentRenameResult.ApplicableActions)
+            {
+                foreach (var error in action.GetErrors())
+                {
+                    errorsDetected = true;
+                    break;
+                }
+            }
+
+            // Check errors before applying changes
+            if (errorsDetected)
+            {
+                string failureMessage = string.Format(CultureInfo.CurrentCulture, VSResources.CannotApplyRename, oldName);
+                _userNotificationServices.ShowWarning(failureMessage);
+                return;
+            }
+
+            if (_languageServiceDone.ContainsKey((oldFilePath, newFileNameWithExtension)))
+            {
+                return;
+            }
+            _languageServiceDone.Add((oldFilePath, newFileNameWithExtension), new SemaphoreSlim(0, 1));
+
+            // Rename the file
+            await base.RenameAsync(context, node, value);
+
+            //Check if there are any symbols that need to be renamed
+            if (documentRenameResult.ApplicableActions.IsEmpty)
+                return;
+
+            // Ask if the user wants to rename the symbol
+            bool userWantsToRenameSymbol = await CheckUserConfirmation(oldName);
+            if (!userWantsToRenameSymbol)
+                return;
+
+            _threadingService.RunAndForget(async () =>
+            {
+                // TODO - implement PublishAsync() to sync with LanguageService
+                // https://github.com/dotnet/project-system/issues/3425)
+                // await _languageService.PublishAsync(treeVersion);
+                //
+                // Because HandleRename() is called after Roslyn, we can set this 
+                // flag when node.FilePath == HandlRename(FilePath)
+                if (_languageServiceDone.TryGetValue((oldFilePath, newFileNameWithExtension), out SemaphoreSlim semLanguegeServiceDone)) {
+                    await semLanguegeServiceDone.WaitAsync();
+                    _languageServiceDone.Remove((node.FilePath, value));
+                }
+
+                // Apply actions and notify other VS features
+                CodeAnalysis.Solution currentSolution = GetCurrentProject().Solution;
+                string renameOperationName = string.Format(CultureInfo.CurrentCulture, VSResources.Renaming_Type_from_0_to_1, oldName, value);
+                (WaitIndicatorResult result, CodeAnalysis.Solution renamedSolution) = _waitService.WaitForAsyncFunctionWithResult(
+                                title: VSResources.Renaming_Type,
+                                message: renameOperationName,
+                                allowCancel: true,
+                                token => documentRenameResult.UpdateSolutionAsync(currentSolution, token));
```

Comment:
If the user cancels, we still try to apply the changes.

Reply:
Which token to use when user cancels ?

Reply:
WaitIndicatorResult returns the result.